handlers/omni/v2/fileformat/csv/reader.go

```diff
@@ -0,0 +1,156 @@
+package omniv2csv
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/antchfx/xpath"
+	"github.com/jf-tech/go-corelib/caches"
+	"github.com/jf-tech/go-corelib/ios"
+	"github.com/jf-tech/go-corelib/maths"
+
+	"github.com/jf-tech/omniparser/idr"
+)
+
+// ErrInvalidHeader indicates the header of the CSV input is corrupted, mismatched, or simply
+// unreadable. This is a fatal, non-continuable error.
+type ErrInvalidHeader string
+
+func (e ErrInvalidHeader) Error() string { return string(e) }
+
+// IsErrInvalidHeader checks if an err is of ErrInvalidHeader type.
+func IsErrInvalidHeader(err error) bool {
+	switch err.(type) {
+	case ErrInvalidHeader:
+		return true
+	default:
+		return false
+	}
+}
+
+type reader struct {
+	inputName     string
+	decl          *fileDecl
+	xpath         *xpath.Expr
+	r             *ios.LineNumReportingCsvReader
+	headerChecked bool
+}
+
+func (r *reader) Read() (*idr.Node, error) {
+	if !r.headerChecked {
+		err := r.checkHeader()
+		r.headerChecked = true
+		if err != nil {
+			return nil, err
+		}
+	}
+read:
+	record, err := r.r.Read()
+	if err == io.EOF {
+		return nil, io.EOF
+	}
+	if err != nil {
+		return nil, r.FmtErr("failed to fetch record: %s", err.Error())
+	}
+	n := r.recordToNode(record)
+	if r.xpath != nil && !idr.MatchAny(n, r.xpath) {
+		goto read
+	}
+	return n, nil
+}
+
+func (r *reader) checkHeader() error {
+	if r.decl.HeaderRowIndex == nil {
+		_ = r.jumpTo(r.decl.DataRowIndex - 1)
```

Comment:
Do we mention anywhere that `HeaderRowIndex`/`DataRowIndex` is 1-based? May better to indicate that if not.

Reply:
Sure will add an comment in the decl. Overall we will have formal usage doc and development doc created a bit later.